substrate/frame/staking/src/delegation.rs

```diff
@@ -0,0 +1,226 @@
+// This file is part of Substrate.
+
+// Copyright (C) Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! An implementation of a delegation system that can be utilised by an off-chain entity, a smart
+//! contract or a runtime module.
+//!
+//! Delegatee: Someone who accepts delegations. An account can set their intention to accept
+//! delegations by calling `accept_delegations`. This account cannot have another role in the
+//! staking system and once set as delegatee, can only stake with their delegated balance, i.e.
+//! cannot use their own free balance to stake. They can also block new delegations by calling
+//! `block_delegations` or remove themselves from being a delegatee by calling 'kill_delegatee' once
+//! all delegations to it are removed.
+//!
+//! Delegator: Someone who delegates their funds to a delegatee. A delegator can delegate their
+//! funds to one and only one delegatee. They also can not be a nominator or validator.
+//!
+//! Reward payouts are always made to another account set by delegatee. This account is a separate
+//! account from delegatee and rewards cannot be restaked automatically. The reward payouts can then
+//! be distributed to delegators by the delegatee via custom strategies.
+//!
+//! Any slashes to a delegatee (which is equivalent to nominator as long as StakingLedger is
```

Comment:
This sounds very much like a pallet that similar to `fast-unstake` can be built externally. Something to consider.

Reply:
I was thinking about this as well to have a `pallet-delegate`. 
- A pallet-delegate that holds funds for an account for a given reason such as for Staking delegation. It maps which account the fund is delegated to.
- Pallet-delegate becomes Currency provider for Staking pallet. When free balance is called for a delegated account, it returns delegated balance else falls back to normal balance.
- When hold funds are called, for delegated funds it just adds some state that it cannot be undelegated until the second lock by the pallet is released.

Would be great to brainstorm this (and also devil's advocate this a bit to not over-engineer it)..